core/src/main/scala/org/apache/spark/rdd/UnionRDD.scala

```diff
@@ -56,14 +58,30 @@ private[spark] class UnionPartition[T: ClassTag](
   }
 }
 
+object UnionRDD {
+  private lazy val partitionEvalThreadPool = new ForkJoinPool(8)
+}
+
 @DeveloperApi
 class UnionRDD[T: ClassTag](
     sc: SparkContext,
     var rdds: Seq[RDD[T]])
   extends RDD[T](sc, Nil) {  // Nil since we implement getDependencies
 
+  // visible for testing
+  private[spark] val isPartitionEvalParallel: Boolean =
+    rdds.length > conf.getInt("spark.rdd.parallelListingThreshold", 10)
+
   override def getPartitions: Array[Partition] = {
-    val array = new Array[Partition](rdds.map(_.partitions.length).sum)
+    val parRDDs = if (isPartitionEvalParallel) {
+      val parArray = rdds.par
+      parArray.tasksupport = new ForkJoinTaskSupport(UnionRDD.partitionEvalThreadPool)
+      parArray
+    } else {
+      rdds
+    }
+    // evaluating parRDDs.map(_.partitions) will execute in parallel and cache partitions
```

Comment:
I don't think this comment adds any value. First of all it's not necessarily true. Second if it's a parallel seq then that's just what Scala does, no surprises there.